src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -80,75 +93,78 @@ class ArrayDeque[A] private[ArrayDeque](
     this
   }
 
-  override def ++=:(elems: TraversableOnce[A]): this.type = {
+  override def ++=:(elems: TraversableOnce[A]) = {
     // Note this is faster than super.++=: because this does not need to convert TraversableOnce to a Traversable
-    if (elems.isEmpty) return this
-    elems.sizeHintIfCheap match {
-      case srcLength if srcLength >= 0 && 2*(srcLength + this.length) >= mask =>
-        val finalLength = srcLength + this.length
-        val array2 = ArrayDeque.alloc(finalLength)
-        elems.copyToArray(array2.asInstanceOf[Array[A]])
-        copySliceToArray(srcStart = 0, dest = array2, destStart = srcLength, maxItems = size)
-        set(array = array2, start = 0, end = finalLength)
-      case _ =>
-        val copy = clone()
-        end = start
-        this ++= elems ++= copy
+    if (elems.nonEmpty) {
+      ArrayDeque.knownSize(elems) match {
+        case srcLength if srcLength >= 0 && 2*(srcLength + this.length) >= mask =>
+          val finalLength = srcLength + this.length
+          val array2 = ArrayDeque.alloc(finalLength)
+          elems.copyToArray(array2.asInstanceOf[Array[A]])
+          copySliceToArray(srcStart = 0, dest = array2, destStart = srcLength, maxItems = size)
+          resetInternal(array = array2, start = 0, end = finalLength)
+        case _ =>
+          val copy = clone()
+          end = start
+          this ++= elems ++= copy
```

Comment:
Actually I tried to implement this and it is 2x slower when array is reversed - but for a surprisingly different reason (nothing to do with indexing is slow). This is because `copySliceToArray` is the most critical section of this data structure and it uses memcpy via System.arrayCopy underneath - when dir is -1 (i.e. array is reversed) there is no way to memcopy in reverse order fast (unless you loop which is ~2x slower than System.arrayCopy).